File: Backtracking/NM2/CS_NM2/NM2.cs
using System;
using System.Collections.Generic;

namespace CS_NM1
{
    static class Constants
    {
        public const int Max = 8 + 1;
    }

    class NM2
    {
        static int n, m;
        static List<int> numList = new List<int>();
        static bool[] checkList = new bool[Constants.Max];

        static void Combination(int cnt, int cur)
        {
            if (cnt == m)
            {
                foreach (var num in numList)
                {
                    Console.Write("{0} ", num);
                }
                Console.WriteLine();
                return;
            }

            for (int i = cur; i <= n; i++)
            {
                if (!checkList[i])
                {
                    numList.Add(i);
                    checkList[i] = true;
                    Combination(cnt + 1, i + 1);
                    numList.RemoveAt(cnt);
                    checkList[i] = false;
                }
            }
        }

        static void Main(string[] args)
        {
            var input = Console.ReadLine().Split();
            n = Convert.ToInt32(input[0]);
            m = Convert.ToInt32(input[1]);

            Combination(0, 1);
        }
    }
}

File: Backtracking/NM2/Python_NM2/NM2.py
import sys

MAX = 8 + 1

n, m = map(int, sys.stdin.readline().split())
num_list = []
check_list = [False for _ in range(MAX)]


def combination(cnt, cur):
    if cnt == m:
        print(*num_list)
        return

    for i in range(cur, n + 1):
        if not check_list[i]:
            num_list.append(i)
            check_list[i] = True
            combination(cnt + 1, i + 1)
            num_list.pop()
            check_list[i] = False


combination(0, 1)
